### DISCUSSION TITLE
Chapter 4: Creating the Application

### DISCUSSION THREAD

**Discussion by jelster** [jelster/space-truckers#125] [Going the Distance Discussion]
Questions, Comments, discussion around the fourth chapter of Going the Distance with Babylon.js

**Comment by AlexanderPolenov** [comment]
Hello Josh,
I recently bought a book (paper version), and now I am on the chapter 4.
I downloaded a code (Ch4) and I have to say that it wasn't properly tested. And I consider several problems to be critical, so it can easily frustrate newcomers.

1) We are getting "reference before assignment" error right away here:
```
this._music = new Sound("titleMusic", titleMusic, scene, () => logger.logInfo("loaded title music"), { autoplay: true, loop: true, volume: 0.5 });
```
It's because the `scene` is defined several lines later.

2) Missing imports for `BABYLON.setAndStartTimer` also lead to app crash immediately.

3) Fade In and Fade Out don't work.
a) Fade Out: 
We call `this._menuContainer.isVisible = false;` before our fading logic starts to work. So, we hide menu before our smooth transition.
I think that the line mentioned above should be inside of the `OnEnded` callback.
b) Fade In:
`_onMenuEnter` is never called. Also, we need to set alpha to 0 right after `menuContainer` creation. Because initially we don't show a menu.

4) `*appStateMachine` generator is not working due to syntax error. Even in playground version: https://playground.babylonjs.com/#EK321G#6
Actually, I really doubt that we should use generator here. Generators is a pretty complex and error-prone feature and should be used only when code really can benefit from it. In our case, we just need to switch couple of variables (previousState and currentState). Why not to do this directly for simplicity?

I understand that some of these problems are fixed in next chapters, but the chapter 4 is still an introduction chapter, and syntax & logic errors can lead to poor understanding and frustration of learners. I believe, this is not what we want.

I created a branch for (Ch4) which fixes some of critical issues mentioned above (also .gitignore IDEs exclusions and Node v17 open-ssl compatibility problem on Windows), but I can't push it here to create a PR, because I don't have access. But it's okay, I think I explained the problems pretty well, so anyone who has access can fix it easily.

**Reply by jelster** [reply]
Hi Alexander, 

Thanks for reading the book and thanks for taking the time to comment with your feedback - it is much appreciated. 

Given the issues you've mentioned encountering, it's understandable why you might feel that the code wasn't properly tested, and I apologize for letting these problems make it into the final branch for this chapter. 

Although I can assure you however that there was plenty of testing at the time it was written, around 2 years ago, this testing was split between code listed in the book, code posted in the source repos branch, as well as code running in Playgrounds. One major challenge in writing the book and application concurrently was making sure that everything is consistent between the different avenues that code is presented, and this early in the book I guess I was still learning how best to do that. 

Still, that's not going to help newcomers or anyone for that matter who are having similar troubles. I've created an Issue for this in the repos to track any work that comes out of this so we can improve the experience for future readers. Let's go through some of the items you raise and recall at the same time that future chapters both build off of and improve the work from previous ones:

> 1. We are getting "reference before assignment" error right away here

Yeah, I can see what you mean by that. Although it shouldn't have been that way to begin with, I think that it I think that this may have slipped through because at that time, `Sound` wasn't checking for/using the `Scene` until after the audio file was loaded, when the scene _would_ be available. Either that or I slipped that line in a last-minute change to a late-night session 😆 This isn't an issue in later commits (ch 5+), fortunately

> 2. Missing imports for BABYLON.setAndStartTimer also lead to app crash immediately

This is something that is resolved in chapter 5 commits. Probably came from bringing in some Playground code.  

> 3. Fade In/Out don't work

Nice catches there, but there's some interesting historical nuance to it as well. In the pre- v5.0 versions of BJS this was originally written against (~ca. early 2021), placing `this._menuContainer._isVisible = false` in the onEnded callback ended up fading the menu while leaving the clicked button visible. IIRC, this was a semi-workaround, because the menuContainer wasn't visible but the clicked button was; setting alpha at the container level adjusted the alpha for children. Does your branch (with code placed more logically) have a fadeout similar to what I desrcribed?

>4. *appStateMachine generator is not working due to syntax error. Even in playground version

I'm not getting any syntax errors with the PG at that link. What's the syntax error and what browser are you using? It's broadly [supported](https://caniuse.com/?search=function*) in most browsers, though I wouldn't recommend Safari. 

> Actually, I really doubt that we should use generator here...

I appreciate the point you make here, however there is a method to my madness 😄 
Later in the book we'll be introduced to the BJS version of CoRoutines, which is where having this knowledge will be useful. I agree that as it is demonstrated in use with as an application state manager seems like overkill, but anything more complicated would have made it more difficult for people unfamiliar with the concepts involved.

Thanks for putting together that PR - I'll take a look at the repository settings to see how to set it up so you can submit it via your fork

**Reply by AlexanderPolenov** [reply]
3. Yes, buttons stay visible. I think, the right way to solve it is to fade out the buttons as well, and we also need to disable 'onHover' behavior for buttons during transition. But using `this._menuContainer._isVisible = false` totally overrides the whole fadeout logic even for background.

4. There is no syntax error that leads to crash. But the logic is not working. If you try to use `moveNextAppState` and `currentState` you will see that there is no actual state change, despite that Logger prints something into console. I saw this also fixed in Ch5 (or later). But it should be working in Ch4, otherwise long explanation of how generators work doesn't bring much understanding, because later the code is broken. I believe this part with generator code snippets should be fixed in the book as well.

Actually, I didn't fork the project I just cloned it locally from https://github.com/jelster/space-truckers directly. I didn't expect that I will need to do a lot of fixes. I think you can apply those fixes yourself in the way that will better match your coding style. And this is possibly not all. There could be more problems after detailed code analysis. I just wrote what I saw right away and considered as critical, because the logic described in the chapter 4 doesn't work.

I understand that many things are fixed in the following chapters, but absolute BabylonJS newcomers may give up, before they will get there.
